src/imperative/imperative_utils.h

```diff
@@ -70,12 +70,14 @@ inline Context GetContext(const nnvm::NodeAttrs& attrs,
   if (ctx.dev_mask() != ctx.dev_type) {
     ctx = Context::Create(ctx.dev_mask(), ctx.dev_id);
   }
-#if !MXNET_USE_CUDA
   if (ctx.dev_mask() == gpu::kDevMask) {
+#if !MXNET_USE_CUDA
     LOG(INFO) << "GPU support is disabled. Compile MXNet with "
               << "USE_CUDA=1 to enable GPU support.";
-  }
+#else
+    mshadow::SetDevice<gpu>(ctx.dev_id);
```

Comment:
It makes no sense to call setdevice for the main thread. This adds overhead.

Reply:
The call is intended to check if the device_id is available. The overhead should be pretty low as it is only setting the dev id corresponding to the thread and there is no synchronization taking place. The other option is to get the cuda device count and check if it the device falls in the range but that would have a similar overhead.

Reply:
Why can't this be handled as an async error in the engine? We can raise the error when user calls asnumpy

Reply:
@piiswrong Where would the async error be raised first ? Currently only exception thrown inside ExecuteOprBlock(during operator exec) is handled. One of the arguments to ExecuteOprBlock is the runtime context and so the stream needs to be set based on device availability before calling ExecuteOprBlock. Therefore, I cannot add this when pushing to engine/handle it as async error in engine.

Reply:
@piiswrong what do you suggest ?

Reply:
you can check it in PushAsync/Sync.

Reply:
@piiswrong Done.